[mlir][EmitC] Add MathToEmitC pass for math function lowering to EmitC

recursion-man opened the pull request:
This commit introduces a new MathToEmitC conversion pass that lowers selected math operations from the Math dialect to the emitc.call_opaque operation in the EmitC dialect.

**Supported Math Operations:**
The following operations are converted:

- math.floor -> emitc.call_opaque<"floor">
- math.round -> emitc.call_opaque<"round">
- math.exp -> emitc.call_opaque<"exp">
- math.cos -> emitc.call_opaque<"cos">
- math.sin -> emitc.call_opaque<"sin">
- math.acos -> emitc.call_opaque<"acos">
- math.asin -> emitc.call_opaque<"asin">
- math.atan2 -> emitc.call_opaque<"atan2">
- math.ceil -> emitc.call_opaque<"ceil">
- math.absf -> emitc.call_opaque<"fabs">
- math.powf -> emitc.call_opaque<"pow">

**Target Language Standards:**
The pass supports targeting different language standards:

- C99: Generates calls with suffixes (e.g., floorf, fabsf) for single-precision floats.
- CPP11: Prepends std:: to functions (e.g., std::floor, std::fabs).

**Design Decisions:**
The pass uses emitc.call_opaque instead of emitc.call to better emulate C-style function overloading.
emitc.call_opaque does not require a unique type signature, making it more suitable for operations like <math.h> functions that may be overloaded for different types.
This design choice ensures compatibility with C/C++ conventions.

recursion-man committed:
[mlir][EmitC] Add MathToEmitC pass for math function lowering to EmitC

This commit introduces a new `MathToEmitC` conversion pass that lowers selected math operations to the `emitc.call_opaque` operation in the EmitC dialect.

The supported math operations include:
- math.floor -> emitc.call_opaque<"floor">
- math.exp -> emitc.call_opaque<"exp">
- math.cos -> emitc.call_opaque<"cos">
- math.sin -> emitc.call_opaque<"sin">
- math.ipowi -> emitc.call_opaque<"pow">

We chose to use `emitc.call_opaque` instead of `emitc.call` to better align with C-style function overloading. Unlike `emitc.call`, which requires unique type signatures, `emitc.call_opaque` allows us to call functions without specifying a unique type-based signature. This flexibility is essential for mimicking function overloading behavior as seen in `<math.h>`.

Additionally, the pass inserts an `emitc.include` operation to generate `#include <math.h>` at the top of the module to ensure the availability of the necessary math functions in the generated code.

This pass enables the use of EmitC as an intermediate layer to generate C/C++ code with opaque calls to standard math functions.

recursion-man commented:
@marbre @TinaAMD @aniragil @simon-camp @cferry-AMD I would greatly appreciate your feedback on this PR whenever you have time. As I don't have write permissions for the repository, I am unable to assign reviewers directly.
Please let me know if any further information is required from my side. Thank you very much for your time and consideration!

cferry-AMD commented on mlir/lib/Conversion/MathToEmitC/MathToEmitC.cpp:
```diff
@@ -0,0 +1,99 @@
+// using function names consistent with those in <math.h>.
+static void populateConvertMathToEmitCPatterns(RewritePatternSet &patterns) {
+  auto *context = patterns.getContext();
+  patterns.insert<LowerToEmitCCallOpaque<math::FloorOp>>(context, "floor");
+  patterns.insert<LowerToEmitCCallOpaque<math::RoundEvenOp>>(context, "rint");
+  patterns.insert<LowerToEmitCCallOpaque<math::ExpOp>>(context, "exp");
+  patterns.insert<LowerToEmitCCallOpaque<math::CosOp>>(context, "cos");
+  patterns.insert<LowerToEmitCCallOpaque<math::SinOp>>(context, "sin");
+  patterns.insert<LowerToEmitCCallOpaque<math::AcosOp>>(context, "acos");
+  patterns.insert<LowerToEmitCCallOpaque<math::AsinOp>>(context, "asin");
+  patterns.insert<LowerToEmitCCallOpaque<math::Atan2Op>>(context, "atan2");
+  patterns.insert<LowerToEmitCCallOpaque<math::CeilOp>>(context, "ceil");
+  patterns.insert<LowerToEmitCCallOpaque<math::AbsFOp>>(context, "fabs");
+  patterns.insert<LowerToEmitCCallOpaque<math::FPowIOp>>(context, "powf");
+  patterns.insert<LowerToEmitCCallOpaque<math::IPowIOp>>(context, "pow");
```
If you have a look at [the page for fabs](https://en.cppreference.com/w/c/numeric/math/fabs) for instance, you'll see that the C implementation (without overloading) bears a different name for each type used. How about a more generic pattern with a predicate:

```cpp
patterns.insert<LowerToEmitCCallOpaque<math::AbsFOp>>(context, "fabsf", [](Type t){ return t == IntegerType::get(t.getContext(), 32); });
patterns.insert<LowerToEmitCCallOpaque<math::AbsFOp>>(context, "fabs", [](Type t){ return t == IntegerType::get(t.getContext(), 64); });
```

marbre replied:
Thanks for addressing this @cferry-AMD. This is something that we definitely should consider. However, I think it depends on what language standard one targets (see https://en.cppreference.com/w/c/numeric/math/fabs and https://en.cppreference.com/w/cpp/numeric/math/fabs). What about an optional switch to specify which standard to target as part of the conversion?

cferry-AMD replied:
How specific to Math functions do you see this pass? The second choice makes the lowering much more generic than I thought: it may apply to any op that could be lowered to opaque library calls, and not just Math dialect ops to `math.h` functions. I like this ability to lower anything to opaque calls with language filters, but this may not be what you intended.

recursion-man replied:
Thank you for the feedback!
I intended this pass to focus on lowering math functions specifically. I accepted the suggestion to use libraries like <cmath> and am also considering the option to add custom math libraries, not just the standard ones, which motivated me to use a map. I can enforce that only math dialect ops will be converted to ensure it remains targeted.

cferry-AMD commented:
Thanks for this PR @recursion-man! I second the approach to go straight to EmitC instead of through a libm detour with `func.call` to allow overloading, and besides this also gives us the opportunity to call functions outside of libm (or go with inline implementations) if, say, the type isn't supported by libm.

marbre reviewed (changes requested):
Thanks for the PR! I think we had a discussion before that this could be nice but seems there was no strong need before.

While I like the approach, I have some concerns regarding non scalar-types. From a quick look it seems
```mlir
%f = math.absf %g : tensor<4xf32>
```
would be converted to valid EmitC and the emitter would translate to C / C++ without an error. I think this needs to be addressed as we cannot in the emitter.

recursion-man committed:
[MLIR][MathToEmitC] Ensure scalar type handling and refactor

This patch ensures that the MathToEmitC pass only converts scalar `FloatType`s, avoiding invalid conversions of non-scalar types like tensors.

- **Validation:** Added checks to convert only scalar types.
- **Refactoring:** Moved implementation to `MathToEmitCPass.cpp` and split headers.
- **Testing:** Added test cases to ensure proper error handling for non-scalar types.

recursion-man reviewed (commented):
1) Added float type check - make sure only scalar float types are handled.
2) Refactoring: moved implementation to MathToEmitCPass.cpp and split headers + use auto-generated constructor

recursion-man committed:
[MLIR][MathToEmitC] Refactor code, add tests for unsupported types, and restrict to f32 only

Refactored code (added newlines and nits).
Added tests to verify behavior with unsupported types.
Now only f32 is supported.
Deleted generation of emitc.include
Changed the conversion to apply at the operation level instead of the module level.

marbre commented on mlir/include/mlir/Conversion/MathToEmitC/MathToEmitC.h:
```diff
@@ -0,0 +1,19 @@
+class RewritePatternSet;
+
+void populateConvertMathToEmitCPatterns(RewritePatternSet &patterns);
+
```
```suggestion
```

recursion-man replied:
Changed

marbre commented on mlir/include/mlir/Conversion/MathToEmitC/MathToEmitC.h:
```diff
@@ -0,0 +1,19 @@
+//===- MathToEmitC.h - Math to EmitC Pass -----------*- C++ -*-===//
```
```suggestion
//===- MathToEmitC.h - Math to EmitC Patterns -------------------*- C++ -*-===//
```

recursion-man replied:
Changed

marbre commented on mlir/include/mlir/Conversion/MathToEmitC/MathToEmitCPass.h:
```diff
@@ -0,0 +1,21 @@
+//===- MathToEmitCPass.h - Math to EmitC Pass -----------------*- C++ -*-===//
```
```suggestion
//===- MathToEmitCPass.h - Math to EmitC Pass -------------------*- C++ -*-===//
```

recursion-man replied:
Done

marbre commented on mlir/include/mlir/Conversion/Passes.td:
```diff
@@ -780,6 +780,22 @@ def ConvertMathToSPIRV : Pass<"convert-math-to-spirv"> {
+def ConvertMathToEmitC  : Pass<"convert-math-to-emitc", "ModuleOp"> {
+  let summary = "Convert some Math operations to EmitC Call_opaque";
+  let description = [{
+    This pass converts supported Math ops to opaque_call ops targeting libc/libm functions.
+    Unlike convert-math-to-funcs pass, this pass uses call_opaque,
+    therefore enables us to overload the same funtion with different argument types
```
Idk the exact line length limit but I think lines should be broken after fewer character. Beside that a suggestion how one could rephrase a little:
```suggestion
    This pass converts supported Math ops to `opaque_call` ops targeting libc/libm
    functions. Unlike convert-math-to-funcs pass, converting to `call_opaque` ops
    allows to overload the same function with different argument types.
```

recursion-man replied:
Refactored

marbre commented on mlir/include/mlir/Conversion/Passes.td:
```diff
@@ -780,6 +780,22 @@ def ConvertMathToSPIRV : Pass<"convert-math-to-spirv"> {
+    Unlike convert-math-to-funcs pass, this pass uses call_opaque,
+    therefore enables us to overload the same funtion with different argument types
+  }];
+  let dependentDialects = ["emitc::EmitCDialect",
+                           "math::MathDialect"
+  ];
```
```suggestion
  let dependentDialects = ["emitc::EmitCDialect"];
```
If I am correct you only need to add the dialect here from that you produce ops not the one you start from.

recursion-man replied:
You were right. I deleted "math::MathDialect"

marbre commented on mlir/lib/Conversion/MathToEmitC/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+  Core
+
+  LINK_LIBS PUBLIC
+  MLIRLLVMCommonConversion
```
```suggestion
```
Is this needed?

recursion-man replied:
No, deleted

marbre commented on mlir/lib/Conversion/MathToEmitC/MathToEmitC.cpp:
```diff
@@ -0,0 +1,58 @@
+//===- MathToEmitC.cpp - Math to EmitC Pass Implementation ----------===//
```
```suggestion
//===- MathToEmitC.cpp - Math to EmitC Patterns -----------------*- C++ -*-===//
```

recursion-man replied:
Right

marbre commented on mlir/lib/Conversion/MathToEmitC/MathToEmitCPass.cpp:
```diff
@@ -0,0 +1,57 @@
+namespace {
+
+//  Replaces Math operations with `emitc.call_opaque` operations.
+struct ConvertMathToEmitCPass
+    : public impl::ConvertMathToEmitCBase<ConvertMathToEmitCPass> {
```
```suggestion
struct ConvertMathToEmitC
    : public impl::ConvertMathToEmitCBase<ConvertMathToEmitC> {
```
`Pass` is used in the old auto-generated registration code and it can simply be omitted.

recursion-man replied:
Deleted!

marbre commented on mlir/lib/Conversion/MathToEmitC/MathToEmitCPass.cpp:
```diff
@@ -0,0 +1,57 @@
+
+} // end anonymous namespace
+
+void ConvertMathToEmitCPass::runOnOperation() {
```
```suggestion
void ConvertMathToEmitC::runOnOperation() {
```

marbre commented on mlir/lib/Conversion/MathToEmitC/MathToEmitCPass.cpp:
```diff
@@ -0,0 +1,57 @@
+
+void ConvertMathToEmitCPass::runOnOperation() {
+  auto moduleOp = getOperation();
+  // Insert #include <math.h> at the beginning of the module
```
I am not sure if we not should this to the user as you would want a different header for C++ here. With that one could also refactor to run the partial conversion on the operation instead of the module.

recursion-man replied:
Ok, deleted and changed the conversion to apply at the operation level instead of the module level.

marbre reviewed (changes requested):
I do not yet have a clear opinion on the two options you proposed so far and only a light tendency to option 1 (don't know if there is real interest in option 2 out there and we mostly implement things we see a clear demand for).

Thanks for iterating on the code and already incorporating my prior suggestions! However, I hope you don't mind that I have some more ;)

recursion-man commented on mlir/lib/Conversion/MathToEmitC/MathToEmitC.cpp:
```diff
@@ -0,0 +1,65 @@
+void mlir::populateConvertMathToEmitCPatterns(RewritePatternSet &patterns) {
+  auto *context = patterns.getContext();
+  patterns.insert<LowerToEmitCCallOpaque<math::FloorOp>>(context, "floor");
+  patterns.insert<LowerToEmitCCallOpaque<math::RoundEvenOp>>(context, "rint");
```
You're right- for now I support only round

recursion-man commented on mlir/lib/Conversion/MathToEmitC/MathToEmitC.cpp:
```diff
@@ -0,0 +1,65 @@
+      !llvm::all_of(actualOp->getResultTypes(),
+                    [](mlir::Type type) { return isa<FloatType>(type); })) {
+    op.emitError("non-float types are not supported");
+    return mlir::failure();
```
Ok, changed the lit test accordingly

recursion-man commented on mlir/lib/Conversion/MathToEmitC/MathToEmitCPass.cpp:
```diff
@@ -0,0 +1,58 @@
+  target.addIllegalOp<math::FloorOp, math::ExpOp, math::RoundEvenOp,
+                      math::CosOp, math::SinOp, math::Atan2Op, math::CeilOp,
+                      math::AcosOp, math::AsinOp, math::AbsFOp, math::PowFOp,
+                      math::FPowIOp, math::IPowIOp>();
```
Right

recursion-man commented on mlir/test/Conversion/MathToEmitC/math-to-emitc.mlir:
```diff
@@ -0,0 +1,133 @@
+// RUN: mlir-opt --split-input-file -convert-math-to-emitc -verify-diagnostics %s | FileCheck %s
```
Added math-to-emitc-failed.mlir which tests invalid types.

recursion-man commented on mlir/test/Conversion/MathToEmitC/math-to-emitc.mlir:
```diff
@@ -0,0 +1,133 @@
+// CHECK:           return
+// CHECK:         }
+func.func @absf_to_call_opaque(%arg0: f32) {
+    %1 = math.absf %arg0 : f32
```
I decided to support only float32 until I decide how to proceed with [this thread](https://github.com/llvm/llvm-project/pull/113799#discussion_r1820643325). I would appreciate you opinion about it.

recursion-man replied:
Added a commit that support both f64 and f32, and updated the lit tests

recursion-man replied:
Looks good. I changed the test accordingly and added one for math.round.

recursion-man commented on mlir/test/Conversion/MathToEmitC/math-to-emitc.mlir:
```diff
@@ -0,0 +1,133 @@
+func.func @test(%arg0 : tensor<4xf32>) -> tensor<4xf32> {
+// expected-error @+2 {{failed to legalize operation 'math.absf' that was explicitly marked illegal}}
+// expected-error @+1 {{non-float types are not supported}}
+  %0 = math.absf %arg0 : tensor<4xf32>
```
I added those to math-to-emitc-failed.mlir

recursion-man commented on mlir/test/Conversion/MathToEmitC/math-to-emitc.mlir:
```diff
@@ -0,0 +1,133 @@
+// expected-error @+1 {{non-float types are not supported}}
+  %0 = math.absf %arg0 : tensor<4xf32>
+  return %0 : tensor<4xf32>
+}
```
Right

recursion-man reviewed (commented):
Refactored the code.
In this commit I decided to support only f32 operands.
In my next commit, I will address language target selection and operand type handling.

recursion-man reviewed (commented):
Refactored and changed the conversion to apply at the operation level instead of the module level.

marbre commented on mlir/lib/Conversion/MathToEmitC/MathToEmitCPass.cpp:
```diff
@@ -0,0 +1,50 @@
+  void runOnOperation() final;
+};
+
+} // end anonymous namespace
```
NIT:
```suggestion
} // namespace
```

recursion-man replied:
Changed, thanks

recursion-man committed:
[MLIR][MathToEmitC] Add support for C and C++ targets with Lit tests

C target: Generates specific callee names based on operand types by appending the appropriate suffix
C++ target: Uses standard library functions with the std:: namespace
Updated LIT tests

recursion-man commented:
Added a new commit that targets both C and C++ and supports f32 and f62.

recursion-man commented:
Hi @marbre, @simon-camp, @cferry-AMD 

I’ve made some updates to the PR, specifically around the decision logic for target support. Now, the changes determine whether we support **C** or **C++** as the target language. Could you please review the latest modifications when you have a chance?

Thank you!

mgehre-amd commented:
Looks good to me

marbre commented on mlir/include/mlir/Dialect/EmitC/IR/EmitCAttributes.td:
```diff
@@ -62,4 +62,13 @@ def EmitC_OpaqueAttr : EmitC_Attr<"Opaque", "opaque"> {
 
 def EmitC_OpaqueOrTypedAttr : AnyAttrOf<[EmitC_OpaqueAttr, TypedAttrInterface]>;
 
+def MathToEmitCLanguageTarget : I32EnumAttr<"MathToEmitCLanguageTarget",
+  "Specifies the language target for generating callees.", [
+    I32EnumAttrCase<"C", 0, "Use C-style function names">,
+    I32EnumAttrCase<"CPP", 1, "Use C++-style function names">
+  ]> {
+  let cppNamespace = "::mlir::emitc";
+}
```
I don't think this should be placed in `EmitCAttributes.td` as it is not strictly belongs to the dialect itself. Scanning over upstream quickly it seems that `ConvertSPIRVToLLVMPass` relies on an Option which uses a custom attribute and not one of the standard types. As another example, `include/mlir/Dialect/ArmSME/Transforms/Passes.td` defines attributes in the passes file itself. I don't know the best solution for this but as mentioned `EmitCAttributes.td`  feels like the wrong place.

aniragil replied:
I wonder if we could take advantage of this use case to introduce a more general attribute attached to the module, e.g.  `EmitCTarget`/`EmitCTargetLanguage`, to be used consistently by lowering passes and the translator where distinct C variants require different lowering, module-level verification or syntax, e.g. 

* Multiple return values are currently lowered to C++ has `std::tuple`, unavailable in other C variants.
* OpenCL-C defines several address spaces, other C variants don't.
* OpenCL-C has native vector types (to some extent), C doesn't.
* C++ has templates, other C variants don't.
* C/C++ support function pointers and recursion, OpenCL-C doesn't.

An alternative to such an attribute might be target-language module ops, e.g. EmitC::C99ModuleOp.

>With the current PR we have "C" and "CPP" as possible targets (I actually would prefer lowercase here). What about >specifying language standard instead? While not having a demand myself right now, I think adding the option to add c23 to e.g. emit fabsd32 could be nice (?). Any (further) opinions on this?

+1
Would be good to enumerate all standard C/C++ versions and let each pass differentiate its lowering as needed.

recursion-man replied:
Thanks for the feedback! 
@marbre, I’m waiting to hear your opinion on @aniragil's suggestion to use a module-level attribute.
As for C23, since decimal types aren’t yet in the builtin types, lowering into them could be challenging. If C23 doesn’t add immediate benefits, maybe we can start with c99 and cpp11. Thoughts?"

marbre replied:
No need to ass C23 from the start but I would prefer a naming (c99, cpp11, ...) that allows to extend if we want to (we would be a little stuck with just `CPP` and `C`).

Let me try to also pull in the other parallel discussion. With regards to @cferry-AMD's comment
>  How specific to Math functions do you see this pass? The second choice makes the lowering much more generic than I thought: it may apply to any op that could be lowered to opaque library calls, and not just Math dialect ops to `math.h` functions. I like this ability to lower anything to opaque calls with language filters, but this may not be what you intended.

and your respond
> Thank you for the feedback! I intended this pass to focus on lowering math functions specifically. I accepted the suggestion to use libraries like and am also considering the option to add custom math libraries, not just the standard ones, which motivated me to use a map. I can enforce that only math dialect ops will be converted to ensure it remains targeted.

I feel like shouldn't put to much into this single PR. We can still use what you have to iterate on it under the assumption that we don't land an approach that introduces drawbacks (for example ops, parameters, ..) that will break with the first refactoring commit and thus break users. Having something that can be consistently used in multiple conversions could be nice, but I currently don't have a concrete idea of how this could look like and lack time to spend much though on it. It would indeed be nice to solve some of the some of issues to which @aniragil pointed but I don't know if we can solve all these with this kind of the proposed attribute. Target module ops like EmitC::C99ModuleOp can be an alternative but are a different concept. Spontaneously, I would prefer to just go a different conversion route (which is what you're doing with the proposed patch) than introducing completely new ops.

recursion-man replied:
@marbre, thanks for the input.
I noticed that **ConvertSPIRVToLLVMPass** relies on a custom Option for **ClientAPI** defined in **SPIRVBase.td**, which is similar to what I’m aiming for here. Additionally, **include/mlir/Dialect/ArmSME/Transforms/Passes.td** defines attributes directly in its passes file but includes **mlir/IR/EnumAttr.td**, going beyond the usual **Conversion/Passes.td** convention to include only **mlir/Pass/PassBase.td**.

At this point, defining the attribute in **EmitCBase.td** or **EmitCAttributes.td** seems like the only alternative I can think of. What are your thoughts?

I also agree with your point on keeping this PR focused without adding too much at once. Further ideas we discussed can be part of future work.

simon-camp replied:
Does this have to be an attribute right now? As we don't have a dedicated op we could attach this to yet, we could start with a `LanguageTarget` enum somewhere in `Dialect/EmitC/Transforms`. Once we introduce a module like op we can give it an inherent attribute to define the target.

Additionally I would strip the `Math` prefix from the name as I expect this to be used in more places in the future as @aniragil mentioned.

recursion-man replied:
Sounds about right to me. I've moved the LanguageTarget enum to Dialect/EmitC/Transforms/Passes.h as you suggested and removed the Math prefix to generalize the enum for broader use.

marbre reviewed (commented):
Thanks for further iterating on this. I am currently quite busy and only have limited time for reviews. So please excuse any (future) delays and don't take is as a lack of interest (it is not!).

With the current PR we have "C" and "CPP" as possible targets (I actually would prefer lowercase here). What about specifying language standard instead? While not having a demand myself right now, I think adding the option to add `c23` to e.g. emit `fabsd32` could be nice (?). Any (further) opinions on this?

simon-camp commented on mlir/lib/Conversion/MathToEmitC/MathToEmitC.cpp:
```diff
@@ -0,0 +1,83 @@
+  if (!llvm::all_of(op->getOperandTypes(), llvm::IsaPred<Float32Type, Float64Type>)||
+      !llvm::all_of(op->getResultTypes(),llvm::IsaPred<Float32Type, Float64Type>))
+    return rewriter.notifyMatchFailure(
+        op.getLoc(), "expected all operands and results to be of type f32");
```
```suggestion
        op.getLoc(), "expected all operands and results to be of type f32 or f64");
```

recursion-man replied:
Right

simon-camp reviewed (changes requested):
Overall this looks good, I just added an alternative suggestion to introducing a new attribute.

On top of that can you please update the list of supported math operations in the description?

recursion-man committed:
[MLIR][MathToEmitC] Add language standard option, create LanguageTarget enum class

- Created LanguageTarget enum class in MathToEmitC.h. An enum that specifies the language target for code generation.

- Added language standard target option (c99, cpp11): allowing users to choose between C99 and C++11.

recursion-man commented on mlir/lib/Conversion/MathToEmitC/MathToEmitCPass.cpp:
```diff
@@ -0,0 +1,52 @@
+  ConversionTarget target(getContext());
+  target.addLegalOp<emitc::CallOpaqueOp>();
+
+  target.addIllegalOp<math::FloorOp, math::ExpOp, math::RoundEvenOp,
```
Good catch!

recursion-man reviewed (commented):
Defined the LanguageTarget enum in Dialect/EmitC/Transforms/Passes.h to specify the language standard target for code generation (currently supporting c99 and cpp11).

recursion-man commented:
Hi @marbre and @simon-camp,

I hope you're doing well. I wanted to follow up on my MathToEmitC PR. I’ve been away for the past couple of months due to personal commitments, but I’m now back and ready to move this forward. Thanks for your patience!

Here’s a quick summary of the changes since we last discussed:

Following @simon-camp’s suggestion, I replaced the _MathToEmitCLanguageTarget_ attribute with a _LanguageTarget_ enum in Dialect/EmitC/Transforms/Passes.h. The enum now supports c99 and cpp11 for better precision.
Based on @marbre’s feedback, I kept the scope focused—avoiding module-level attributes or new ops for now.

I’d appreciate it if you could take another look and let me know if there’s anything else I should address.
Thanks again for your input and support!

marbre commented on mlir/include/mlir/Conversion/MathToEmitC/MathToEmitC.h:
```diff
@@ -0,0 +1,21 @@
+//===- MathToEmitC.h - Math to EmitCPatterns -------------------*- C++ -*-===//
```
NIT:
```suggestion
//===- MathToEmitC.h - Math to EmitCPatterns --------------------*- C++ -*-===//
```

marbre commented on mlir/include/mlir/Conversion/MathToEmitC/MathToEmitC.h:
```diff
@@ -0,0 +1,21 @@
+#ifndef MLIR_CONVERSION_MATHTOEMITC_MATHTOEMITC_H
+#define MLIR_CONVERSION_MATHTOEMITC_MATHTOEMITC_H
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Dialect/EmitC/Transforms/Passes.h"
```
I assume this include made was added as the enum is defined in that header. Actually, the header is only included in the two new `mlir/include/mlir/Conversion/MathToEmitC/{MathToEmitC,MathToEmitCPass}.h` header files. Maybe there is a better place to define the enum, `Transforms/Passes.h` seems not the best place. @simon-camp any suggestion?

simon-camp replied:
I would define it right here for now. If/when this get's used by other passes it can move to `Dialect/EmitC/IR/EmitC.h`

recursion-man replied:
Hi @marbre and @simon-camp,

I’ve updated the PR and made the changes as you suggested. Let me know if there’s anything else to address.

simon-camp reviewed (approved):
LGTM, just found two errors in comments.

recursion-man commented on mlir/include/mlir/Conversion/MathToEmitC/MathToEmitC.h:
```diff
@@ -0,0 +1,25 @@
+//===- MathToEmitC.h - Math to EmitCPatterns --------------------*- C++ -*-===//
```
Thanks, fixed

simon-camp reviewed (approved):
Thanks, looks great!

marbre commented on mlir/include/mlir/Conversion/Passes.td:
```diff
@@ -780,6 +780,28 @@ def ConvertMathToSPIRV : Pass<"convert-math-to-spirv"> {
+def ConvertMathToEmitC  : Pass<"convert-math-to-emitc"> {
+  let summary = "Convert some Math operations to EmitC Call_opaque";
+  let description = [{
+    This pass converts supported Math ops to `opaque_call` ops targeting libc/libm
```
```suggestion
    This pass converts supported Math ops to `call_opaque` ops targeting libc/libm
```

marbre commented on mlir/include/mlir/Conversion/Passes.td:
```diff
@@ -780,6 +780,28 @@ def ConvertMathToSPIRV : Pass<"convert-math-to-spirv"> {
+//===----------------------------------------------------------------------===//
+
+def ConvertMathToEmitC  : Pass<"convert-math-to-emitc"> {
+  let summary = "Convert some Math operations to EmitC Call_opaque";
```
```suggestion
  let summary = "Convert some Math operations to EmitC call_opaque operations";
```

marbre commented on mlir/lib/Conversion/MathToEmitC/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+  MLIREmitCDialect
+  MLIRMathDialect
+  MLIRPass
+  MLIRTransforms
```
```suggestion
  MLIRTransformUtils
```
The header file `"mlir/Transforms/DialectConversion.h"` corresponds to `mlir/lib/Transforms/Utils/DialectConversion.cpp` which is part of the MLIRTransformUtils target, see https://github.com/llvm/llvm-project/blob/7dd34baf5505d689161c3a8678322a394d7a2929/mlir/lib/Transforms/Utils/CMakeLists.txt#L5.

recursion-man replied:
Thanks, @marbre. Appreciate your thorough review!

marbre reviewed (changes requested):
Nearly all good, just some minor remarks before this ready to merge :)

marbre reviewed (approved):
Thanks for your work!

marbre commented:
Before we can merge, can you please update the PR description to reflect the current state? The PR description will become the body of the commit message when squashing. Furthermore, according to the Developer Policy, can you please disable that your email address is kept private?

https://llvm.org/docs/DeveloperPolicy.html#making-and-submitting-a-patch
> The LLVM project uses email to communicate to contributors outside of the GitHub platform about their past contributions. Primarily, our buildbot infrastructure uses emails to contact contributors about build and test failures. Therefore, the LLVM community requires contributors to have a public email address associated with their GitHub commits, so please ensure that “Keep my email addresses private” is disabled in your [account settings](https://github.com/settings/emails).

marbre commented:
Seems there are also some formatting changes necessary. I didn't realized that the CI needed approval to fully run.

recursion-man committed:
[mlir][EmitC] Apply formatting fixes

recursion-man commented:
Hi @marbre  and @simon-camp,

Thanks again for your reviews and support! All checks have passed, and the PR is ready to merge. Could you approve the pending workflow and merge it?

Appreciate it!

marbre commented:
Merged, thanks again for your patch!
